gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+        case WMP_errOutOfSequence: return "OutOfSequence:";
+        case WMP_errNotInitialized: return "NotInitialized";
+        case WMP_errMustBeMultipleOf16LinesUntilLastCall: return "MustBeMultipleOf16LinesUntilLastCall";
+        case WMP_errPlanarAlphaBandedEncRequiresTempFile: return "PlanarAlphaBandedEncRequiresTempFile";
+        case WMP_errAlphaModeCannotBeTranscoded: return "AlphaModeCannotBeTranscoded";
+        case WMP_errIncorrectCodecSubVersion: return "IncorrectCodecSubVersion";
+        default: return "Unknown";
+        }
+    }
+};
+
+/************************************************************************/
+/* ==================================================================== */
+/*                         JPEGXRDecoder                                */
+/* ==================================================================== */
+/************************************************************************/
+
+struct JPEGXRDecoder final
+{
+    JPEGXRDecoder()
+        : pCodecFactory(nullptr), pImageDecode(nullptr)
+    {
+    }
+
+    ~JPEGXRDecoder()
+    {
+        if (pImageDecode)
+            pImageDecode->Release(&pImageDecode);
+        pImageDecode = nullptr;
+
+        if (pCodecFactory)
+            pCodecFactory->Release(&pCodecFactory);
+        pCodecFactory = nullptr;
+    }
+
+    void Initialize(char const* pszFilename)
+    {
+        CPLAssert(!pCodecFactory);
+        CPLAssert(!pImageDecode);
+        ERR nErr = WMP_errFail;
+
+        /* -------------------------------------------------------------------- */
+        /*      Create decoder                                                  */
+        /* -------------------------------------------------------------------- */
+        nErr = PKCreateCodecFactory(&pCodecFactory, WMP_SDK_VERSION);
+        if (nErr != WMP_errSuccess || !pCodecFactory)
+            throw JPEGXRException(nErr);
+
+        nErr = pCodecFactory->CreateDecoderFromFile(pszFilename, &pImageDecode);
+        if (nErr != WMP_errSuccess || !pImageDecode)
+            throw JPEGXRException(nErr);
+
+        /* -------------------------------------------------------------------- */
+        /* Self-check understanding of jxrlib internals and data consistency.   */
+        /* -------------------------------------------------------------------- */
+#ifndef NDEBUG
+        PKPixelInfo pi;
+        GetPixelInfo(pi);
+
+        CPLAssert(pi.cChannel == pi.uSamplePerPixel);
+        CPLAssert(pImageDecode->WMP.wmiI.cfColorFormat == pi.cfColorFormat);
+        CPLAssert(pImageDecode->WMP.wmiI.bRGB == ((pi.grBit & PK_pixfmtBGR) == 0));
+        CPLAssert(pImageDecode->WMP.wmiI.cBitsPerUnit == pi.cbitUnit);
+        CPLAssert(pImageDecode->WMP.wmiI.cBitsPerUnit == pi.uSamplePerPixel * pi.uBitsPerSample);
+#endif
+
+        if (nErr != WMP_errSuccess)
+            throw JPEGXRException(nErr);
+    }
+
+    void GetBytesPerPixel(int& nBytePerPixel) const
```

Comment:
@rouault Yes. Those are remains after exception-free version, with return error codes, that I used to have.